demos/mantle_convection/adaptive_base_case/adaptive_base_case.py

```diff
@@ -0,0 +1,362 @@
+    T_init, u_init, p_init - initial conditions, or last solution from previous mesh
+                             we interpolate these onto the current mesh
+
+    returns:
+    time, T, u, p - time, temperature, velocity and pressure at the end of the timesteps
+    steady_state_reached - have we reached steady state during the timesteps?
+    """
+
+    # Set up function spaces - currently using the bilinear Q2Q1 element pair:
+    V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+    W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+    Q = FunctionSpace(mesh, "CG", 1)  # Temperature function space (scalar)
+    Z = MixedFunctionSpace([V, W])  # Mixed function space.
+
+    z = Function(Z, name='Solution')  # A field over the mixed function space Z.
+
+    # Output function space information:
+    log("Number of Velocity DOF:", V.dim())
+    log("Number of Pressure DOF:", W.dim())
+    log("Number of Velocity and Pressure DOF:", V.dim()+W.dim())
+    log("Number of Temperature DOF:", Q.dim())
+
+    # Set up temperature field and extract velocity and pressure from "mixed" function z
+    T = Function(Q, name="Temperature")
+    u, p = z.subfunctions
+
+    # Interpolate temperature, velocity and pressure. In the first call this will
+    # simply interpolate the provided UFL expresssions for the initial conditions.
+    # In subsequent calls this will perform cross-mesh interpolation from the
+    # solutions at the previous mesh, to the current mesh.
+    T.interpolate(T_init)
+    u.interpolate(u_init)
+    p.interpolate(p_init)
+
+    t_adapt = TimestepAdaptor(delta_t, u, V, maximum_timestep=0.1, increase_tolerance=1.5)
+
+    # Stokes related constants (note that since these are included in UFL, they
+    # are wrapped inside Constant):
+    Ra = Constant(1e5)  # Rayleigh number
+    approximation = BoussinesqApproximation(Ra)
+
+    steady_state_tolerance = 1e-9
+
+    # Nullspaces and near-nullspaces:
+    Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=False)
+
+    # Next rename for output:
+    u.rename("Velocity")
+    p.rename("Pressure")
+
+    mesh.cartesian = True
+    boundary = get_boundary_ids(mesh)
+    gd = GeodynamicalDiagnostics(z, T, boundary.bottom, boundary.top)
+
+    temp_bcs = {
+        boundary.bottom: {'T': 1.0},
+        boundary.top: {'T': 0.0},
+    }
+
+    stokes_bcs = {
+        boundary.bottom: {'uy': 0},
+        boundary.top: {'uy': 0},
+        boundary.left: {'ux': 0},
+        boundary.right: {'ux': 0},
+    }
+
+    energy_solver = EnergySolver(T, u, approximation, delta_t, BackwardEuler, bcs=temp_bcs)
+    Told = energy_solver.T_old
+    Ttheta = 0.5*T + 0.5*Told
+    Told.assign(T)
+    stokes_solver = StokesSolver(z, approximation, Ttheta, bcs=stokes_bcs,
+                                 nullspace=Z_nullspace, transpose_nullspace=Z_nullspace,
+                                 solver_parameters="iterative")
```

Comment:
This part needs explaining. Why is `Ttheta` required? Is `Told.assign(T)` required?